notebooks/04_model_select_and_optimize/nni_ncf.ipynb

```diff
@@ -0,0 +1,762 @@
+       " 'parameter_source': 'algorithm',\n",
+       " 'parameters': {'n_factors': 12, 'learning_rate': 0.0023365527461525885},\n",
+       " 'parameter_index': 0}"
+      ]
+     },
+     "execution_count": 13,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "best_params_ncf"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 5. Baseline Model\n",
+    "\n",
+    "Although we hope that the additional effort of utilizing an AutoML framework like NNI for hyperparameter tuning will lead to better results, we should also draw comparisons using our baseline model (our model trained with its default hyperparameters). This allows us to precisely understand what performance benefits NNI is or isn't providing."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 15,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "data = NCFDataset(train, validation, seed=DEFAULT_SEED)\n",
+    "model = NCF(\n",
+    "    n_users=data.n_users, \n",
+    "    n_items=data.n_items,\n",
+    "    model_type=\"NeuMF\",\n",
+    "    n_factors=4,\n",
+    "    layer_sizes=[16,8,4],\n",
+    "    n_epochs=NUM_EPOCHS,\n",
+    "    learning_rate=1e-3,  \n",
+    "    verbose=True,\n",
+    "    seed=DEFAULT_SEED\n",
+    ")\n",
+    "model.fit(data)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 16,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def compute_test_results(model, train, test):\n",
+    "    test_results = {}\n",
+    "    \n",
+    "    # Rating Metrics\n",
+    "    predictions = [[row.userID, row.itemID, model.predict(row.userID, row.itemID)]\n",
+    "           for (_, row) in test.iterrows()]\n",
+    "\n",
+    "    predictions = pd.DataFrame(predictions, columns=['userID', 'itemID', 'prediction'])\n",
+    "    predictions = predictions.astype({'userID': 'int64', 'itemID': 'int64', 'prediction': 'float64'})\n",
+    "\n",
+    "    for metric in RATING_METRICS:\n",
+    "        test_results[metric] = eval(metric)(test, predictions)\n",
+    "        \n",
+    "    # Ranking Metrics\n",
+    "    users, items, preds = [], [], []\n",
+    "    item = list(train.itemID.unique())\n",
+    "    for user in train.userID.unique():\n",
+    "        user = [user] * len(item) \n",
+    "        users.extend(user)\n",
+    "        items.extend(item)\n",
+    "        preds.extend(list(model.predict(user, item, is_list=True)))\n",
+    "\n",
+    "    all_predictions = pd.DataFrame(data={\"userID\": users, \"itemID\": items, \"prediction\": preds})\n",
+    "\n",
+    "    merged = pd.merge(train, all_predictions, on=[\"userID\", \"itemID\"], how=\"outer\")\n",
+    "    all_predictions = merged[merged.rating.isnull()].drop('rating', axis=1)\n",
+    "\n",
+    "    for metric in RANKING_METRICS:\n",
+    "        test_results[metric] = eval(metric)(test, all_predictions, col_prediction='prediction', k=K)\n",
```

Comment:
Hopefully this will build up the metrics iteratively, and reuse calculations from previously steps. This for loop could end up running in O(log(n)) time, and not O(n) if calucaltions are reused.

Reply:
this function was moved out of the notebook and I agree that it can be improved. There is a huge bottleneck in this function, we would need to review it in depth if we have more time